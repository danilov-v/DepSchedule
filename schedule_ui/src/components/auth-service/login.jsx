import React from "react";
import PropTypes from "prop-types";
import { Button, Form, FormGroup, FormFeedback, Input } from "reactstrap";
import { useForm } from "helpers/hooks/useForm";

import "./login.scss";

const defaultFormData = {
  login: "",
  password: "",
};

const validateForm = values => {
  let errors = {};

  if (!values.login || values.login.trim().length < 3)
    errors["login"] = "Имя пользовтеля должно содержать минимум 3 символа";

  if (!values.password || values.password.trim().length < 5)
    errors["password"] = "Пароль должен содержать минимум 5 символов";

  return errors;
};

export const Login = ({ login }) => {
  const onFormSubmit = async () => {
    try {
      await login(values);
    } catch (e) {
      console.log(e);
      setErrors({ auth: "Неверное имя или пароль" });
    }
  };

  const {
    onChange,
    onSubmit,
    errors,
    errorsShown,
    setErrors,
    values,
  } = useForm(onFormSubmit, defaultFormData, validateForm);

  const handleInputChange = ({ target }) =>
    onChange({ [target["name"]]: target.value });

  return (
    <div className="h-100 w-100 text-dark emblem-background">
      <Form className="login-form">
        <h2 className="text-center mb-5">Вход в Cистему</h2>
        <FormFeedback>{errors["auth"]}</FormFeedback>
        <FormGroup>
          <Input
            type="text"
            rows="6"
            name="login"
            placeholder="Имя пользователя"
            id="note"
            value={values.login}
            onChange={handleInputChange}
            invalid={errorsShown && (!!errors["auth"] || !!errors["login"])}
          />
          <FormFeedback>{errors["login"]}</FormFeedback>
        </FormGroup>
        <FormGroup>
          <Input
            type="password"
            rows="6"
            name="password"
            placeholder="Пароль"
            id="password"
            value={values.password}
            onChange={handleInputChange}
            invalid={errorsShown && (!!errors["auth"] || !!errors["password"])}
          />
          <FormFeedback>{errors["auth"] || errors["password"]}</FormFeedback>
        </FormGroup>
        <Button type="submit" color="primary" block onClick={onSubmit}>
          Войти
        </Button>
      </Form>
    </div>
  );
};

Login.propTypes = {
  login: PropTypes.func,
};
