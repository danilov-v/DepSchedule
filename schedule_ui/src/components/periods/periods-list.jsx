import React from "react";
import PropTypes from "prop-types";
import { Table, Button } from "reactstrap";
import { MANAGE_PERIODS } from "constants/permishions";
import { checkPermission } from "utils/permishions";

export function PeriodsList({
  periods,
  onPeriodEdit,
  onPeriodRemove,
  userRole,
}) {
  const isManageAble = checkPermission(userRole, MANAGE_PERIODS);

  const renderRow = (period, index) => (
    <tr key={period.periodId}>
      <th scope="row">{index}</th>
      <td>{period.name}</td>
      <td>{period.startDate}</td>
      <td>{period.endDate}</td>
      <td hidden={!isManageAble}>
        <Button onClick={onPeriodEdit.bind(null, period)} color="warning">
          Изменить
        </Button>
        <Button close onClick={onPeriodRemove.bind(null, period.periodId)} />
      </td>
    </tr>
  );
  return (
    <div className="event-type-list">
      <Table hover>
        <thead>
          <tr>
            <th>#</th>
            <th>Название</th>
            <th>Начало</th>
            <th>Конец</th>
            <th hidden={!isManageAble}></th>
          </tr>
        </thead>
        <tbody>{periods.map(renderRow)}</tbody>
      </Table>
    </div>
  );
}

PeriodsList.propTypes = {
  periods: PropTypes.arrayOf(
    PropTypes.shape({
      periodId: PropTypes.number,
      name: PropTypes.string,
      startDate: PropTypes.string,
      endDate: PropTypes.string,
    })
  ),
  onPeriodEdit: PropTypes.func.isRequired,
  onPeriodRemove: PropTypes.func.isRequired,
  userRole: PropTypes.string,
};

PeriodsList.defaultProps = {
  periods: [],
};
