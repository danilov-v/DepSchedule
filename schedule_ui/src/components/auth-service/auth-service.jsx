import React, { useState, useContext, createContext, useEffect } from "react";
import { Redirect } from "react-router-dom";
import { isString, isEmpty, get } from "lodash";
import { signIn, signOut } from "helpers/api";
import { NotificationManager } from "helpers/notification-manager";
import { FAILED_LOGIN } from "constants/notifications";
import { Login } from "./login";

export const DEFAULT_AUTH_DATA = {
  authBody: JSON.parse(window.localStorage.getItem("authBody")) || {},
};
const AuthServiceContext = createContext({});

export const useAuth = () => useContext(AuthServiceContext);

export const AuthServiceProvider = ({ children }) => {
  const [authBody, setAuthBody] = useState(DEFAULT_AUTH_DATA.authBody);

  useEffect(() => {
    window.localStorage.setItem(
      "authBody",
      isString(authBody) ? authBody : JSON.stringify(authBody)
    );
  }, [authBody]);

  const login = async userCred => {
    try {
      const data = await signIn(userCred);
      setAuthBody(data);
    } catch (e) {
      NotificationManager.fire(FAILED_LOGIN);
      throw e;
    }
  };

  const logout = async () => {
    setAuthBody({});

    try {
      await signOut();
    } catch (e) {
      console.group("failed server logout with error: ");
      console.log(e);
      console.groupEnd();
    }
  };

  const getRole = () => get(authBody, "role");

  const defaultContext = {
    authBody,
    getRole,
    login,
    logout,
  };

  return (
    <AuthServiceContext.Provider value={defaultContext} children={children} />
  );
};

export const AuthPage = () => {
  const { authBody, login } = useAuth();

  return isEmpty(authBody) ? <Login login={login} /> : <Redirect to="/" />;
};
