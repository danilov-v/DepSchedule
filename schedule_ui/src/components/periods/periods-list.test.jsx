import React from "react";
import render from "enzyme/render";
import { AuthServiceProvider } from "components/auth-service/auth-service";
import { PeriodsList } from "./periods-list";

const props = {
  onPeriodEdit: () => {},
  onPeriodRemove: () => {},
  periods: [
    {
      periodId: 0,
      name: "test",
      startDate: "2019-01-01",
      endDate: "2020-01-01",
    },
  ],
};

describe("periods-list.jsx", () => {
  describe("<PeriodsList />", () => {
    test("should render and match snapshot", () => {
      const component = render(
        <AuthServiceProvider>
          <PeriodsList {...props} />
        </AuthServiceProvider>
      );

      expect(component).toMatchSnapshot();
    });
  });
});
