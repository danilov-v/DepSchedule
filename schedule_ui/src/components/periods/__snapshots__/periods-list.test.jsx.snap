// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`periods-list.jsx <PeriodsList /> should render and match snapshot 1`] = `
<div
  class="event-type-list"
>
  <table
    class="table table-hover"
  >
    <thead>
      <tr>
        <th>
          #
        </th>
        <th>
          Название
        </th>
        <th>
          Начало
        </th>
        <th>
          Конец
        </th>
        <th
          hidden=""
        />
      </tr>
    </thead>
    <tbody>
      <tr>
        <th
          scope="row"
        >
          0
        </th>
        <td>
          test
        </td>
        <td>
          2019-01-01
        </td>
        <td>
          2020-01-01
        </td>
        <td
          hidden=""
        >
          <button
            class="btn btn-warning"
            type="button"
          >
            Изменить
          </button>
          <button
            aria-label="Close"
            class="close"
            type="button"
          >
            <span
              aria-hidden="true"
            >
              ×
            </span>
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
